transport/ssh.go

```diff
@@ -0,0 +1,180 @@
+	homedir, err := homeDir()
+	if err != nil {
+		return nil, err
+	}
+
+	return knownhosts.New(filepath.Join(homedir, ".ssh", "known_hosts"))
+}
+
+func sshConnect(host, port, user string) (*ssh.Client, error) {
+	connID := user + "@" + host + ":" + port
+	// This is a single-use cache, so pass in an empty category.
+	obj, err := connectionCache.GetOrUpdate("", connID, expires, true, func() (interface{}, error) {
+		agent, err := newAgent()
+		if err != nil {
+			return nil, fmt.Errorf("Failed to find config from SSH_AUTH_SOCK: %v", err)
+		}
+
+		hostKeyCallback, err := getHostKeyCallback()
+		if err != nil {
+			return nil, fmt.Errorf("Loading SSH known hosts file: %v", err)
+		}
+
+		sshConfig := &ssh.ClientConfig{
+			User:            user,
+			Auth:            []ssh.AuthMethod{agent},
+			HostKeyCallback: hostKeyCallback,
+		}
+
+		return ssh.Dial("tcp", host+":"+port, sshConfig)
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	return obj.(*ssh.Client), nil
+}
+
+// Identity identifies how to connect to a target.
+type Identity struct {
+	Host, User string
+}
+
+// ExecSSH executes against a target via SSH. It will look up port, user, and other configuration
+// by exact hostname match from default SSH config files. Identity can be used to override the
+// default user.
+//
+// Command will be shell quoted into a single invocation string. Ignores the TTY option.
+//
+// Lots of SSH configuration is currently omitted, such as global known hosts files, finding known
+// hosts from the config, identity file from config... pretty much everything but port and user
+// from config as enumerated in https://github.com/kevinburke/ssh_config/blob/0.5/validators.go.
+func ExecSSH(ctx context.Context, id Identity, cmd []string, opts plugin.ExecOptions) (plugin.ExecResult, error) {
+	var result plugin.ExecResult
+
+	// find port, username, etc from .ssh/config
+	port, err := ssh_config.GetStrict(id.Host, "Port")
+	if err != nil {
+		return result, err
+	}
+
+	user := id.User
+	if user == "" {
+		if user, err = ssh_config.GetStrict(id.Host, "User"); err != nil {
+			return result, err
+		}
+	}
+
+	if user == "" {
+		user = "root"
+	}
+
+	connection, err := sshConnect(id.Host, port, user)
+	if err != nil {
+		return result, fmt.Errorf("Failed to connect: %s", err)
+	}
+
+	// Run command via session
+	session, err := connection.NewSession()
+	if err != nil {
+		return result, fmt.Errorf("Failed to create session: %s", err)
+	}
+
+	outputch, stdout, stderr := plugin.CreateExecOutputStreams(ctx)
+	session.Stdin, session.Stdout, session.Stderr = opts.Stdin, stdout, stderr
+
+	cmdStr := shellquote.Join(cmd...)
+	if err := session.Start(cmdStr); err != nil {
+		return result, err
+	}
+
+	// Wait for session to complete and stash result. This is done asynchronously because the
+	// exitcode callback won't be called until stdout/stderr are closed, but they're not closed by
+	// session.Start.
+	var exitErr error
+
+	go func() {
```

Comment:
This pattern is becoming common enough that it might be worth including some sort of cancellation function in `ExecResult`. That way as a plugin author, I don't have to worry about contexts, which is Go-specific.

Reply:
We could require the post-condition that if Exec returns without an error, then the command was successfully started. That would make it easier to reason about how this stuff works.

Reply:
That makes some sense.